### DISCUSSION TITLE
[Enhancement] [popconfirm] popconfirm 基于 Promise 的异步关闭

### DISCUSSION THREAD

**Discussion by MarTuPoke** [element-plus/element-plus#12248] [Feature Request]
Feature type: **`Enhancement`**

## Details

- Enhanced component name: `popconfirm`

## Description

基于promise的关闭，参考antd https://ant.design/components/popconfirm-cn#components-popconfirm-demo-promise

### Additional comments

_(empty)_

**Comment by gimjin** [comment]
貌似通过 [beforeClose](https://element-plus.org/en-US/component/message-box.html#customization) 已支持异步关闭，只是方案不同。

**Reply by MarTuPoke** [reply]
popconfirm 组件

**Reply by MarTuPoke** [reply]
不是 message-box

**Comment by gimjin** [comment]
的确 antd 和 naive-ui 确实是支持 promise。不过他们是用prop来作为事件来处理的，整个 element-plus 的组件代码中并没有看到用 prop作为事件的案例。 el-upload 有一些 promise 勾子，但是帮助文档中是 prop 来出现的。

vue 早不支持[有返回值的 emit](https://github.com/vuejs/core/commit/55566e8f520eee8a07b85221174989c47c443c35) ，如果带着 done回调函数不够优雅感觉像回调地狱模式，如果用promise代码优雅，但是整个代码里面没有先例。

不知道该怎么贡献 :(

**Reply by MarTuPoke** [reply]
我有一个思路，还是通过props传递进去，然后再确定按钮的触发事件中，判断是否有props传递进来，如果没有，还是正常逻辑，如果有，则进行执行promise，执行完再继续后续

**Reply by MarTuPoke** [reply]
```js
const handleConfirm = async () => {
    if(props.confirmFn){
        await props.confirmFn()
    }
    // TOOD ...
}
```

**Reply by MarTuPoke** [reply]
这个需求是很有必要的，因为 确认按钮 一点击 就关闭，其实还有一些逻辑在里面，比如说，拒绝删除，删除中等状态其实没有体现出来，不一定要通过promise来改造事件的触发，只是说 能不能在关闭前，将一些控制权交给用户，其实更像是beforeClose

**Reply by gimjin** [reply]
是的， naive-ui里面几乎看不到 emit，帮助文档里面也没有 event。 不过 element-plus 并没有那么做，文档里面清晰地区别了 props 和 events。 如果代码里面用 prop 定义，然后帮助文档里面用 event 提示肯定是不行的。 

目前看上去只能把里面的 [hidePoper](https://github.com/element-plus/element-plus/blob/dev/packages/components/popconfirm/src/popconfirm.vue#L86) 通过参数传递出来，但是这样写很不好看。

按照 naive-ui 那样写 vuejs emit 基本就废了，按照当前element-plus的写法不太优雅，左右为难还是闭着眼睛捏着鼻子先用 [ElPopover](https://element-plus.org/en-US/component/popover.html#nested-operation) 吧。

**Reply by MarTuPoke** [reply]
el-popover 就把控制权完全掌握在自己手上了
